File: zanorpc/spend.go
package zanorpc

import (
	"bytes"
	"context"
	"crypto/rand"
	"fmt"

	"github.com/KarpelesLab/zanolib"
	"github.com/KarpelesLab/zanolib/zanobase"
)

// RingSize is the total ring size for ZC inputs on mainnet
// (CURRENCY_HF4_MANDATORY_DECOY_SET_SIZE = 15 decoys + 1 real).
const RingSize = 16

// txHardforkID is the hardfork id carried by current (post-HF5) mainnet
// transactions; observed as 5 on all current v3 txs.
const txHardforkID = 5

// SweepTo builds, signs and (optionally) broadcasts a transaction that sends all
// of the given deposits back to recipient, deducting `fee` (native atomic units)
// from the native amount. Confidential-asset deposits are forwarded in full.
// Decoy rings are fetched per input via getrandom_outs3.
//
// It returns the signed transaction, its raw serialized bytes, and (if
// broadcast) the daemon status. The wallet must hold the spend secret.
func (c *Client) SweepTo(ctx context.Context, wallet *zanolib.Wallet, deposits []*Deposit, recipient string, fee uint64, broadcast bool) (*zanolib.FinalizedTx, []byte, string, error) {
	if len(deposits) == 0 {
		return nil, nil, "", fmt.Errorf("no deposits to sweep")
	}
	addr, err := zanolib.ParseAddress(recipient)
	if err != nil {
		return nil, nil, "", fmt.Errorf("recipient address: %w", err)
	}

	native := zanolib.NativeCoinAssetId()

	// Build inputs (fetch a decoy ring for each) and tally per-asset totals.
	var inputs []*zanolib.TransferInput
	totals := map[string]uint64{} // asset key -> total amount
	assetOrder := []string{}      // preserve first-seen order
	assetBytes := map[string]zanobase.Value256{}
	for _, d := range deposits {
		ring, err := c.GetRingForOutput(ctx, d.GlobalIndex, RingSize)
		if err != nil {
			return nil, nil, "", fmt.Errorf("decoys for gindex %d: %w", d.GlobalIndex, err)
		}
		members := make([]zanolib.RingMember, len(ring))
		for i, o := range ring {
			members[i] = zanolib.RingMember{
				GlobalIndex:      o.GlobalIndex,
				StealthAddress:   o.StealthAddress,
				ConcealingPoint:  o.ConcealingPoint,
				AmountCommitment: o.AmountCommitment,
				BlindedAssetId:   o.BlindedAssetId,
			}
		}
		inputs = append(inputs, &zanolib.TransferInput{
			Amount:              d.Out.Amount,
			AssetId:             d.Out.AssetId,
			AmountBlindingMask:  d.Out.AmountBlindingMask,
			AssetIdBlindingMask: d.Out.AssetIdBlindingMask,
			RealOutTxKey:        d.TxPubKey,
			RealOutIndex:        uint64(d.Out.OutputIndex),
			RealGlobalIndex:     d.GlobalIndex,
			Ring:                members,
		})
		k := string(d.Out.AssetId[:])
		if _, ok := totals[k]; !ok {
			assetOrder = append(assetOrder, k)
			assetBytes[k] = d.Out.AssetId
		}
		totals[k] += d.Out.Amount
	}

	// One destination per asset, all to the recipient; fee comes off native.
	var dests []*zanolib.TransferDest
	for _, k := range assetOrder {
		amount := totals[k]
		if k == string(native[:]) {
			if amount < fee {
				return nil, nil, "", fmt.Errorf("native total %d is less than fee %d", amount, fee)
			}
			amount -= fee
		}
		if amount == 0 {
			continue
		}
		dests = append(dests, &zanolib.TransferDest{
			Address: addr,
			AssetId: assetBytes[k],
			Amount:  amount,
		})
	}

	signed, err := wallet.BuildTransfer(rand.Reader, inputs, dests, zanobase.TransactionVersionPostHF5, txHardforkID)
	if err != nil {
		return nil, nil, "", fmt.Errorf("build transfer: %w", err)
	}

	var buf bytes.Buffer
	if err := zanobase.Serialize(&buf, signed.Tx); err != nil {
		return signed, nil, "", fmt.Errorf("serialize tx: %w", err)
	}
	raw := buf.Bytes()

	if !broadcast {
		return signed, raw, "", nil
	}
	status, err := c.SendRawTx(ctx, raw)
	if err != nil {
		return signed, raw, "", err
	}
	return signed, raw, status, nil
}

File: src/rpc/spend.rs
//! Sweeping received deposits back out, with decoy rings fetched per input.

use super::client::Client;
use super::scanner::Deposit;
use crate::base::{EpeeWrite, TRANSACTION_VERSION_POST_HF5, Value256};
use crate::error::{Error, Result};
use crate::rng::OsRng;
use crate::{
    Address, FinalizedTx, RingMember, TransferDest, TransferInput, Wallet, native_coin_asset_id,
};
use std::collections::BTreeMap;

/// Total ring size for confidential inputs on mainnet
/// (`CURRENCY_HF4_MANDATORY_DECOY_SET_SIZE` = 15 decoys + 1 real).
pub const RING_SIZE: usize = 16;

/// The hardfork id carried by current (post-HF5) mainnet transactions.
pub const TX_HARDFORK_ID: u64 = 5;

/// The result of [`Client::sweep_to`].
pub struct SweepResult {
    /// The signed transaction.
    pub signed: FinalizedTx,
    /// Its serialized bytes.
    pub raw: Vec<u8>,
    /// The daemon's status string, if it was broadcast.
    pub status: Option<String>,
}

impl Client {
    /// Builds, signs and optionally broadcasts a transaction sending all of
    /// `deposits` to `recipient`, deducting `fee` (native atomic units) from the
    /// native amount. Confidential-asset deposits are forwarded in full.
    ///
    /// Decoy rings are fetched per input via `getrandom_outs3`. The wallet must
    /// hold the spend secret.
    pub fn sweep_to(
        &self,
        wallet: &Wallet,
        deposits: &[Deposit],
        recipient: &str,
        fee: u64,
        broadcast: bool,
    ) -> Result<SweepResult> {
        if deposits.is_empty() {
            return Err(Error::msg("no deposits to sweep"));
        }
        let addr = Address::parse(recipient).map_err(|e| crate::err!("recipient address: {e}"))?;
        let native = native_coin_asset_id();

        // Build inputs (fetching a decoy ring for each) and tally per-asset totals.
        let mut inputs = Vec::with_capacity(deposits.len());
        let mut totals: BTreeMap<Value256, u64> = BTreeMap::new();
        let mut asset_order: Vec<Value256> = Vec::new();

        for d in deposits {
            let ring = self
                .get_ring_for_output(d.global_index, RING_SIZE)
                .map_err(|e| crate::err!("decoys for gindex {}: {e}", d.global_index))?;
            let members: Vec<RingMember> = ring
                .iter()
                .map(|o| RingMember {
                    global_index: o.global_index,
                    stealth_address: o.stealth_address,
                    concealing_point: o.concealing_point,
                    amount_commitment: o.amount_commitment,
                    blinded_asset_id: o.blinded_asset_id,
                })
                .collect();
            inputs.push(TransferInput {
                amount: d.out.amount,
                asset_id: d.out.asset_id,
                amount_blinding_mask: d.out.amount_blinding_mask.clone(),
                asset_id_blinding_mask: d.out.asset_id_blinding_mask.clone(),
                real_out_tx_key: d.tx_pub_key,
                real_out_index: d.out.output_index as u64,
                real_global_index: d.global_index,
                ring: members,
            });
            if !totals.contains_key(&d.out.asset_id) {
                asset_order.push(d.out.asset_id);
            }
            *totals.entry(d.out.asset_id).or_default() += d.out.amount;
        }

        // One destination per asset, all to the recipient; the fee comes off native.
        let mut dests = Vec::new();
        for asset in &asset_order {
            let mut amount = totals[asset];
            if *asset == native {
                if amount < fee {
                    return Err(crate::err!("native total {amount} is less than fee {fee}"));
                }
                amount -= fee;
            }
            if amount == 0 {
                continue;
            }
            dests.push(TransferDest {
                address: addr.clone(),
                asset_id: *asset,
                amount,
            });
        }

        let signed = wallet
            .build_transfer(
                &mut OsRng,
                &inputs,
                &dests,
                TRANSACTION_VERSION_POST_HF5,
                TX_HARDFORK_ID,
            )
            .map_err(|e| crate::err!("build transfer: {e}"))?;

        let raw = signed.tx.to_epee_bytes();
        if !broadcast {
            return Ok(SweepResult {
                signed,
                raw,
                status: None,
            });
        }
        let status = self.send_raw_tx(&raw)?;
        Ok(SweepResult {
            signed,
            raw,
            status: Some(status),
        })
    }
}
